src/plugins/share/public/persistable_state/persistable_state_service.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { CoreSetup, CoreStart, Plugin } from 'src/core/public';
+import { identity } from 'lodash';
+import { PersistableStateDefinition, PersistableStateContract } from './types';
+
+export interface PersistableStateSetup {
+  register: <Id extends string>(
+    id: Id,
+    definition?: PersistableStateDefinition<Id>
+  ) => Promise<void>;
+}
+
+export interface PersistableStateStart {
+  get: <Id extends string>(definitionId: Id) => Promise<PersistableStateContract<Id>>;
+}
+
+export class PersistableStateService
+  implements Plugin<PersistableStateSetup, PersistableStateStart> {
+  private definitions = new Map();
+
+  public setup(core: CoreSetup): PersistableStateSetup {
+    return {
+      register: async (id, definition = {}) => {
+        const { extractReferences, injectReferences, migrate } = definition;
+
+        this.definitions.set(id, {
+          id,
+          extractReferences: extractReferences || ((s: unknown) => [s, []]),
+          injectReferences: injectReferences || identity,
+          migrate: migrate || identity,
+        });
+      },
+    };
+  }
+
+  public start(core: CoreStart): PersistableStateStart {
+    return {
+      get: async id => {
+        const definition = this.definitions.get(id);
+        if (!definition) {
+          throw new Error(
+            i18n.translate('share.persistableState.errors.noStateWithId', {
+              defaultMessage: 'No persistable state found with id {id}',
+              values: { id },
+            })
+          );
```

Comment:
Added error handling here, but then later realized in the last PR we discussed possibly just returning `undefined`.

I have no strong feelings either way.

Reply:
if we throw then everywhere we'll use the service we need to do something like:

```
let newState;
try {
  newState = share.persistableState.get('visualization').migrate(mystate);
} catch {}
// if there is no state found we still want to continue, there is just no migration to do
```

where if we return undefined we need to:
```
const newState = share.persistableState.get('visualization') ? share.persistableState.get('visualization').migrate(state) : state;
```

thats why i suggested always returning, but in case nothig is registered returning default implementation of persistable state utils, so you can then just:


```
 const newState = share.persistableState.get('visualization').migrate(state)